Fixes for function and variable declarations

Daniel-Cortez opened the pull request:
**What this PR does / why we need it**:

This PR does the following:
* Allows to use a combination of `static stock`/`stock static` in new-style function declarations (see #622).
* Fixes assertion failure when defining `static` variables with a name starting with `@` (see #623).
* Makes the compiler detect the use of specifier `static` on variables whose name starts with `@` as an invalid combination of class specifiers (see #625).
* Addresses problems with function class specifiers having an incorrect effect or no effect at all in function declarations (see #621, #624) by introducing the following rules:
  * Function definition must have all class specifiers that were introduced in all previous declarations of the function.
```Pawn
forward public Func1(); // Class specifier "public" is introduced; it will be required in the definition.
Func1(); // OK (previously introduced class specifiers are only mandatory in function definitions,
         // not declarations).
Func1(){} // Error (missing class specifier "public").

static Func2();  // Class specifier "static" is introduced.
forward stock Func2(); // Class specifier "stock" is introduced; now the definition is required to have
                       // both keywords "static" and "stock".
stock static Func2(){} // OK (both class specifiers "static" and "stock" are in place).
```
  * * After the function is defined, subsequent "forward" re-declarations can't introduce any new specifiers.
```Pawn
static stock Func(){}
forward public Func(); // Error (function wasn't defined as "public").
```
* Fixes a bug with the return value tag and array size being expected before the function class specifiers in new-style function declarations (see #635).
* Fixes a bug with keyword `__pragma` being expected twice in function/variable declarations starting with keywords `public`, `static` and `stock`.

**Which issue(s) this PR fixes**:

Fixes #621, #622, #623, #624, #625, #635

**What kind of pull this is**:

* [x] A Bug Fix
* [x] A New Feature
* [ ] Some repository meta (documentation, etc)
* [ ] Other

**Additional Documentation**:

Daniel-Cortez committed:
Allow a combination of `static` and `stock` specifiers in new-style function declarations

Daniel-Cortez committed:
Forbid the use of `static` and `stock` specifiers on functions/variables whose name starts with `@`

Daniel-Cortez committed:
Require specifiers `public`, `static` and `stock` in function definitions and repeated declarations if previous declaration(s) had those specifiers

Daniel-Cortez committed:
Unset function class specifier flags at the end of compiler pass, so on the next pass "forward" declarations won't be required to have those specifiers at first

Daniel-Cortez committed:
Don't accept new class specifiers in "forward" re-declarations if the function has already been defined

Daniel-Cortez committed:
Add a separate flag to memoize the use of "public" specifier in function declarations/definitions.
Also rename "uSTATIC" into "uDECLSTATIC".

Daniel-Cortez committed:
Add tests

Daniel-Cortez committed:
Rename the "stock" argument into "fstock" to better match the naming of other arguments

Daniel-Cortez committed:
Simplify a few lines of code in `funcstub()`

Daniel-Cortez committed:
funcstub(): Expect the function tag and the return array size after the class specifiers and `__pragma`

Daniel-Cortez committed:
Update tests

Daniel-Cortez committed:
Move the code for handling operator `__pragma` out of `newfunc()`

This fixes a bug with `__pragma` being expected twice in certain cases.

Daniel-Cortez committed:
Expect `__pragma` before the tag in function/variable declarations

Daniel-Cortez committed:
Move the code for handling `__pragma` in prefix form out of `newfunc()`

This fixes another bug with `__pragma` being expected twice in variable declarations starting with specifiers `public`, `static` and `stock`.

Daniel-Cortez committed:
Update tests

Daniel-Cortez committed:
Only require function definitions to have all specifiers introduced in forward declarations

Daniel-Cortez committed:
Update tests

Y-Less commented:
Those missing specifier rules are causing YSI to stop building.  I know I usually say that breaking YSI isn't a blocker for new compiler versions because it does a lot of silly things, but I'm not entirely sure this is a silly thing.  If a forward declaration has `stock`, it seems more sensible to me that the `stock` rule is added to the function, regardless of what the definition then says.  Leads to some interesting things like:

```pawn
MAKE_STOCK(func);

func()
{
}
```

Daniel-Cortez committed:
Do not require specifier `stock`, even if it was previously introduced in one of function declarations

Daniel-Cortez committed:
Allow introduction of specifier `stock` in post-definition forward declarations

Example:
  Func(){}
  forward stock Func();

Daniel-Cortez committed:
Update tests

Daniel-Cortez commented:
> ```pawn
> MAKE_STOCK(func);
> 
> func()
> {
> }
> ```
Hmm... haven't seen such non-trivial use of keyword `stock` before, but I guess it can be useful. I'll see if I can relax those new rules for that specifier.

Daniel-Cortez commented:
To clarify why I added this new requirement for class specifiers in function implementations, my main concern was that in 3.10.10 (without this PR) class specifiers either had no effect in function declarations (see #624) or could even make the compiler generate invalid code (#621), so this new requirement was supposed to point users at the parts of their code that could cause problems in compiler versions up to 3.10.10.

@Y-Less I already removed that rule for specifier `stock`, as it couldn't impact generated code (it only inhibits warning 204 in case the function is unused), but do you want me to remove that requirement for `static` and `public` as well?

Daniel-Cortez committed:
Don't accept new class specifiers in "old-style" re-declarations if the function has already been defined

Daniel-Cortez committed:
Update tests

Y-Less commented:
I'm actually not sure what the most sensible result is here.  It seems there are equally good arguments for "all specifiers are additive", so this declares a `static stock` function:

```pawn
forward static Func();
forward stock Func();
Func()
{
}
```

"all specifiers must build", so only this is valid:

```pawn
forward static Func();
forward static stock Func();
static stock Func()
{
}
```

"only the definition has specifiers":

```pawn
forward Func();
forward Func();
static stock Func()
{
}
```

In that I don't think any one of them is obviously more "correct" than the others.  Having said that, I'd vote for the first one as it gives the most options.  Regardless of which is chosen the rule should be applied evenly to `stock`, `static`, and `public`, not just having one special-cased as you did for `stock`.

Y-Less commented:
There is another option - remove old-style declarations, but that's a bigger change.

Y-Less commented:
Or rahter that's a breaking change.

Daniel-Cortez committed:
Allow function class specifiers to be "additive", so they aren't required in following function re-declarations or definition

Daniel-Cortez committed:
Clarify that specifier `stock` can be accepted even after the function is already defined

Daniel-Cortez committed:
Detect invalid combinations of "static" and "public" when each specifier is introduced in its own function declaration or definition, e.g.:
```
forward static Func();
forward public Func();
```

Daniel-Cortez committed:
Don't repeat error 042 twice when a function is defined as `static` and its name starts with `@`

Daniel-Cortez committed:
Update tests
